New version of Radius2john.py

k4amos opened the pull request:
`radius2john.py` didn't seem to be able to work properly : 

- The attack method is described  [here](http://www.untruth.org/~josh/security/radius/radius-auth.html). The previous version of `radius2john.py` implemented the _3.1 Response Authenticator Based Shared Secret Attack_. The attacker can pre-compute the MD5 state for (Code+ID+Length+RequestAuth+Attributes). However, the value of the attributes was forgotten in the code, the calculated hash could not be broken if there were attributes.
- The _3.3 User-Password Attribute Based Shared Secret Attack_ was not implemented

For these two reasons, I propose a new version of radius2john.py much closer to the perl version _radius2john.pl_

![image](https://github.com/openwall/john/assets/70712826/1e94c5f0-3ba8-4249-8d93-9f03a94b7029)

k4amos committed:
New version of Radius2john.py

solardiz commented:
Thank you @k4amos! For reference, the previous version was contributed via #3621.

I'd appreciate review by @MaximeGoyette @exploide @kholia.

Is this revised script for Python 3 and/or 2? The current shebang lines implies it supports both, which is great if so, but I'm not sure is the case.

k4amos committed:
shebang python3 + processing RADIUS accounting messages on port 1813 + try / except scapy

k4amos commented:
Hello @solardiz,

I added the python3 shebang and a try / except block to display a message if scapy is not installed.

RADIUS accounting messages are on port 1813, I had forgotten to consider them, but I corrected that. The shared-secret can be cracked using these methods, as demonstrated by running the code on the first sample provided by Maxime Goyette https://github.com/openwall/john/pull/3621#issuecomment-458752330 

I realise that the perl version has also forgotten to consider accounting messages, do I create an issue for the perl version ?

exploide commented on run/radius2john.py:
```diff
@@ -1,68 +1,139 @@
+# "3.1 Response Authenticator Based Shared Secret Attack"
+
+# For attack 3.3 :
+# we try authentications using a known password, and sniff the radius packets to a pcpap file.
```
typo pcap

exploide commented on run/radius2john.py:
```diff
@@ -1,68 +1,139 @@
+# This script reads the pcap file, matches radius responses with the corresponding all_requests,
+# and dumps md5 and salt as needed.
+
+import scapy.all as scapy
```
There is a scapy import here as well as in the `if __name__ == "__main__"` block. The latter is guarded with a try-catch, but this first import here will crash the script way before it hits the main block. Please move the try-except import up here.

exploide commented on run/radius2john.py:
```diff
@@ -1,68 +1,139 @@
+# For attack 3.3 :
+# we try authentications using a known password, and sniff the radius packets to a pcpap file.
+# This script reads access-request in the pcap file, and dumps the md5(RA+secret) and RA, in a john-friendly format.
+# The password must be always the same, be less then 16 bytes long, and entered in the $PASSWORD variable below.
```
As far as I understood, users need to touch the code for attack 3.3 in order to set the `PASSWORD` and `LOGINS` variables. I think it would be much nicer to have commandline options for that. The [argparse](https://docs.python.org/3/library/argparse.html) module comes in handy here. I can help with this if needed.

exploide commented on run/radius2john.py:
```diff
@@ -1,68 +1,139 @@
+
+        if udp_layer.dport in [1812, 1813] or udp_layer.sport in [1812, 1813]: 
+            radius_data = bytes(udp_layer.payload)
+            process_radius(ip_layer, radius_data, bytes(udp_layer.payload))
```
you set `radius_data = bytes(udp_layer.payload)`, then you pass `radius_data` as well as `bytes(udp_layer.payload)` as parameters to the `process_radius` function. This is two times the same data.

exploide commented on run/radius2john.py:
```diff
@@ -1,68 +1,139 @@
+    if radius_packet.code in [1, 4]:  # Access-Request, Accounting-Request
+
+        user_name, user_password = None, None
+        for _ in range(len(radius_packet.attributes)):
```
This can be simplified to `for attr in radius_packet.attributes:` and then directly accessing `attr.name` and `attr.value`. In Python `for` loops are usually meant as for-each loops. Iterating over indices is only done when really necessary.

exploide commented on run/radius2john.py:
```diff
@@ -1,68 +1,139 @@
+            if radius_packet.attributes[_].name == "User-Password":
+                user_password = radius_packet.attributes[_].value.decode("utf-8")
+
+        if user_name in VALID_LOGIN and user_password:
```
Is it possible that the extracted `user_password` is an empty string and would that still be a valid case? If yes, then the condition needs to be `and user_password is not None` because the empty string also evaluates to `False` in an `if`.

exploide commented on run/radius2john.py:
```diff
@@ -1,68 +1,139 @@
-        if Radius in packet:
-            src_ip = packet[IP].src
-            dst_ip = packet[IP].dst
+def dump_access_request(ip, login, ra, hashed):  # 3.3 attack
```
unused argument `login`

exploide commented on run/radius2john.py:
```diff
@@ -1,68 +1,139 @@
+        print(
+            "Scapy seems to be missing, run 'pip install --user scapy' to install it"
+        )
+        exit(1)
```
using `sys.exit(1)` is recommended over `exit(1)`

exploide reviewed (changes requested):
I had a look on the script. Thanks for the contribution. :)

I left a few comments and suggestions. I only read and linted the script for know, I haven't actually executed it yet. I assume we only have the test data available in https://github.com/openwall/john/pull/3621#issuecomment-458752330 ? Nothing in the john-samples repo unfortunately.

If further questions arise or when I could help with code changes, please note that I'm away over the weekend.

solardiz commented:
> I added the python3 shebang

In other words, you're saying the script is now Python 3 only? Was the previous one working with either Python version? If so, that's a minor regression (but not a blocker).

> I realise that the perl version has also forgotten to consider accounting messages, do I create an issue for the perl version ?

Yes, or ideally a pull request right away. Thank you!

solardiz commented:
> Nothing in the john-samples repo unfortunately.

BTW, @k4amos @MaximeGoyette we welcome contributions via pull requests to https://github.com/openwall/john-samples so that it's easier and more reliable to regression-test future versions.

k4amos committed:
code cleaning and small improvements

k4amos commented:
Hi @solardiz and @solardiz, I hope I've corrected everything correctly. With the samples, you can test the code : the previous version of the code only worked with the first 2 samples, whereas this new version works with all 5 : 

previous version of the code : 
![image](https://github.com/openwall/john/assets/70712826/e7b0d15c-76ff-4a1c-bda9-fbdcf96c2ca8)

new version of the code : (by default, I no longer return a single hash)
![image](https://github.com/openwall/john/assets/70712826/6e56e3e1-13d9-442d-8825-2693fd69ecff)

You can test the 3.3 attack for example with the sample n°5 :
![image](https://github.com/openwall/john/assets/70712826/e8a96b10-21f0-4d20-99bd-f85efdf22689)
![image](https://github.com/openwall/john/assets/70712826/fbf67fcf-85a8-4cd2-aa8b-c4e3757ed0a9)

k4amos commented:
The previous version of the code seems to work well with both python2 and python3 @solardiz , this is a regression on this point, I have no solution to keep this compatibility 🫤

solardiz commented:
> Hi @solardiz and @solardiz, I hope I've corrected everything correctly.

Hi @k4amos and @k4amos. Thank you. For fix-up commits within a PR, we prefer that the original commit be amended and force-pushed, instead of adding more commits. Also, the commit titles should start with `radius2john.py: `, although if the only commit is `New version of radius2john.py` that is also fine.

In this case, you may squash the commits, or you may just tell me to use GitHub's squash-and-merge (which we normally have disabled to avoid inadvertent use, but I can briefly re-enable it for merging this one PR).

I haven't reviewed your actual changes. I hope @exploide will.

Also thank you for the sample files and for the test examples. It would be even better if you post such examples as plain text copy-paste, rather than as screenshots, so they would also be in email.

solardiz commented:
> The previous version of the code seems to work well with both python2 and python3 @solardiz , this is a regression on this point, I have no solution to keep this compatibility 🫤

OK. Does the Perl script provide the same functionality? If so, I suppose it would be the one to use on older systems that don't have Python 3 installed.

Edit: Oh, I suppose this (from an earlier comment) implies the Perl version isn't yet as functional:

> I realise that the perl version has also forgotten to consider accounting messages

exploide commented on run/radius2john.py:
```diff
@@ -1,68 +1,163 @@
+                    user_hash = attr.value
+
+            if user_hash is not None:
+                if (args['login'] is not None and user_name == args['login']) or args['login'] is None:
```
I think this can be simplified to `if args['login'] is None or user_name == args['login']`

exploide commented on run/radius2john.py:
```diff
@@ -1,68 +1,163 @@
-                accounting_request_packets[(src_ip, dst_ip)] = packet
-            elif packet[Radius].code == 5 and not (dst_ip, src_ip) in accounting_response_packets:
-                accounting_response_packets[(dst_ip, src_ip)] = packet
+    parser.add_argument('-f', '--file', type=str, required=True, nargs='+')
```
the `-f` option conflicts with the your hand-written usage message (at multiple places) that the basic usage is `./radius2john.py <pcap files>`. Either adapt the usage message, or better, use a positional argument without option flag for the files.

exploide commented on run/radius2john.py:
```diff
@@ -1,68 +1,163 @@
 
-            salt = code + identifier + length + authenticator
+    parsed_args = parser.parse_args()
+    args = vars(parsed_args)
```
I don't understand why this is necessary. You could just access something like `parsed_args.login`. But it's not important.

exploide commented on run/radius2john.py:
```diff
@@ -1,68 +1,163 @@
 
-            hash_content = hexlify(packet[Radius].authenticator).decode('utf-8')
-            hash_type = 1009 if len(salt) <= 16 else 1017
+    if args["login"] is not None and args["password"] is None:
```
Did I understand correctly, that you always need both, login and password for attack 3.3? If yes, then also a check for password is not None but login is would be required, wouldn't it?

k4amos replied:
I don't necessarily need the login for the 3.3 attack: if there's only one client, there's no point in checking that it's the right one (because that's necessarily the case), but in all cases, I need the password. That's why if I have a login but no password, I generate an error. 

```bash
user@computer:~/Documents$ python3 john/run/radius2john.py -f john-samples/Radius/radius-sample-5.pcap -l bob 
You must specify the password used by the client for the '3.3 User-Password Attribute Based Shared Secret Attack'
Basic Usage: ./radius2john.py -f <pcap files>
You can specify '-h' to display help
```


On the other hand, if I just have just the password but no login, the code can work : 
```bash
user@computer:~/Documents$ python3 john/run/radius2john.py -f john-samples/Radius/radius-sample-5.pcap -p ThisIsThePassword
194.247.5.25:$dynamic_1008$9df88bcafd6b2550fcf2c62c87dd8619$HEX$1da7522f1df80483e73573c72b7d9114
194.247.5.25:$dynamic_1017$4fb969d9dc9d7aee6f58132e0b9662cd$HEX$021300201da7522f1df80483e73573c72b7d9114120c48656c6c6f2c20626f62
```

k4amos replied:
I'm going to add this explanation to the code

exploide reviewed (changes requested):
Thanks for improving the PR. I did another review round. Much less comments now, and less important. Please have a look.

Concerning python2 compatibility: Personally I don't care tbh. If we require python2 compatibility than we cannot use modern language features, like f-strings, need multiple switches and workarounds and increase the maintenance overhead in general. A user however, it not forced to execute the script on an ancient system but can just copy the file to a modern platform. I think lowering the developer and maintenance effort is more important. Python2 is end of life for years. But @solardiz has to decide this. (If you ask me, I would even dump the perl script since this one here has all features the perl script has plus more.)

k4amos committed:
fix unclear explanation

k4amos commented:
If it's possible to make a GitHub's squash-and-merge @solardiz instead of editing the commits already made, that would be perfect 😃

solardiz commented:
> If it's possible to make a GitHub's squash-and-merge @solardiz instead of editing the commits already made, that would be perfect 😃

OK, I'll plan on doing that. Does this look ready, @exploide?

exploide reviewed (approved):
Yes, I guess this is ready now.

solardiz commented:
This is now merged. However, I made the mistake of not letting CI run on this before merging - my thinking was we don't test Python scripts anyway - but we do have the whitespace errors test for all text files, and this test failed. I've now made a follow-up commit fixing the whitespace errors.

@k4amos Please take a look at how this looks in the tree now. Thank you!
